### ISSUE TITLE
`enumerable.map` raise wrong number of arguments only when Ruby script is executed from Gradle using JDK 21

### ISSUE THREAD

**Issue by andsel** [jruby/jruby#8061]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc): `9.4.5.0`
* Operating system and platform (e.g. `uname -a`): <
```
Darwin andreas-MacBook-Pro-2.local 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:55:06 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6020 arm64
```

Other relevant info you may wish to add:
* Installed or activated gems
* Application/framework version (e.g. Rails, Sinatra)
* Environment variables

**Expected Behavior**

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.
When running the simple script using JDK 21
```ruby
hash_set = Java::JavaUtil::LinkedHashSet.new
hash_set << 1
hash_set << 2
hash_set << 3
res = hash_set.map {|v| v % 2}
```
within JRuby's `ScriptingContainer.runScriptlet` should create a new enumerable without throwing error when it's instantiated inside an embedding Java application and also when it's used from Gradle.

* Provide an executable Ruby script or a link to an example repository.
https://github.com/andsel/test_jdk21_jruby

**Actual Behavior**

* Describe or show the actual behavior.
*With JDK 21 the above test runs as expected  in an 
[embedding Java application](https://github.com/andsel/test_jdk21_jruby/blob/6fd243f53a39775b368af498fd0f8985bb87ce09/src/main/java/org/dna/App.java#L20) but fails with 
```
(ArgumentError) wrong number of arguments (given 1, expected 0)
``` 
if the same script is executed form a `ScriptingContainer` container [instantiated inside a Gradle script](https://github.com/andsel/test_jdk21_jruby/blob/6fd243f53a39775b368af498fd0f8985bb87ce09/build.gradle#L48).
* Provide text or screen capture showing the behavior.

**Comment by headius** [comment]
This seems like a case where we're not overriding a Java `map` with the Ruby `map`. The block you pass then gets converted to a Java argument, and the method being called takes none.

However I can't see anything in LinkedHashSet or its parent classes that implements `map`. Hmmm.

**Comment by headius** [comment]
If you get back to this before me, you could set property `jruby.backtrace.style=full` and provide the output. It will show where we're raising this error from in JRuby.

**Comment by andsel** [comment]
Hi @headius thanks for suggestion. I tried running with
```sh
./gradlew -Djruby.backtrace.style=full boom
```

and verified from Gradle script that property is set:
```
Running with System props: [java.specification.version:21, sun.jnu.encoding:UTF-8, sun.arch.data.model:64, java.vendor.url:https://adoptium.net/, jruby.backtrace.style:full,....
```

But no backtrace was printed on console :-(

**Comment by headius** [comment]
Ok I managed to run your example and found the problem. This method is defined on `LinkedHashSet` as of Java 21 (not there in 17):

```java
    @SuppressWarnings("unchecked")
    LinkedHashMap<E, Object> map() {
        return (LinkedHashMap<E, Object>) map;
    }
```

This is interfering with the Ruby `map` method because we include that into java.util.Collection and that comes higher in the hierarchy. However this method is only package-visible, so I don't think we should be binding it in the first place.

**Comment by headius** [comment]
Yeah looks like our logic must be allowing package-private methods to get bound for some reason. I'll try to figure out why.

**Comment by kares** [comment]
Believe binding package-private was intentional at some point, I think for user-controlled packages it *made* some sense.
There used to be an access check in place that would prevent binding methods from packages of a different module...

**Comment by headius** [comment]
There's something odd going on here. When I run the following script at the command line, it works fine:

```
lhs = java.util.LinkedHashSet.new
lhs.map { }
```

The logic in JRuby only permits through `public` methods or methods that can be safely made accessible (module is open, etc). I suspect something in the Gradle version is making that package open, or somehow making this method accessible before we get to it, but I have no hard evidence of that yet.

**Comment by headius** [comment]
Ok it's not Gradle causing this, it's running JRuby from classpath rather than as a module.

When JRuby is run at a command line in a normal installation, our script loads it as the `org.jruby.dist` module, and we then open up a few JDK modules to get access to file descriptors etc. We do not open java.util, and so this package-private `map` method does not get `setAccessible` nor bound into the proxy class.

When loaded as a normal classpath entry or URLClassLoader, JRuby is not registered as a module. As current as Java 21, JDK internals can be opened to the unnamed module, which causes backport9 to do so, set this `map` method accessible, and return `true` to JRuby so it proceeds to bind the method.

If it is possible to have Gradle load the JRuby jar as a module, that would fix this behavior; we would no longer bind the `map` method. I could not find any documentation that showed how to do this.

We may also want to reconsider the JRuby behavior for precisely this sort of case. Protected and package-private methods are not intended to be consumed by external code, and 90% of JRuby use cases will not need (or want) access to them. Where we do need to access them is from Ruby classes that extend Java classes or implement Java interfaces and that should gain access to protected methods, and that is still an open project to take on.

The current `setAccessible` check is a blunt tool, and our heuristic for applying it seems at the very least to be flawed.

It's also arguable that package-private and private should never be made accessible through normal Java proxy class setup, since even child classes should not be able to access them. If that's what you want, you can dig through the java.lang.Class yourself, or we could provide a mechanism in the proxy classes to forcibly bind specific non-visible methods. I don't believe we should do it automatically anymore (except for protected access from child classes).

@kares @enebo I propose we should do the following:

* Narrow the current logic to never try to `setAccessible` package-private methods. It already rejects private methods.
* Possibly extend this to protected methods. As it stands now, such methods will only be accessible and bound when running JRuby as a standalone jar on the classpath, so most users would not notice this change.
* If we extend this to protected methods, we'll need to figure out how to handle the child class case (finally!) Should they be bound as Ruby `protected`? Not bound at all, and we add logic to discover them when called from a child class? It's not obvious how to proceed.

**Comment by headius** [comment]
I [discussed](https://matrix.to/#/!vyEDBdbmzqApWaugdd:matrix.org/$170621802629375wVrze:matrix.org?via=matrix.org) this with @enebo on Matrix and we came to the following compromise:

* When running on JDK releases prior to Java 17 (which finally shut off illegal accesses) we will maintain the current behavior: protected and package-private methods will be `setAccessible` if already opened or if running JRuby in the unnamed module.
* When running on JDK releases from Java 17 forward, we will no longer attempt to `setAccessible` package-private methods, since they would not be accessible to any normal Java code outside that package. That will fix this issue.
* We will not change this behavior for protected methods until we have a suitable workaround for child-class access. Currently the only workaround for that is to explicitly open up the module, and omitting protected method binding altogether would break that workaround.

**Comment by headius** [comment]
Ok, a few minutes after we had decided how to proceed I discovered that I was wrong: this **is** Gradle's fault:

https://github.com/gradle/gradle/blob/f796b166368355a4c62fbbfb3157bd29c3f2a25c/platforms/core-runtime/base-services/src/main/java/org/gradle/internal/jvm/JpmsConfiguration.java#L35

They are setting these packages to be open to all unnamed modules, which includes JRuby in this case. When I run the following example outside of Gradle, it succeeds because the package-private `map` does not get bound:

```
$ java -Djruby.home=`pwd` -jar lib/jruby.jar -e 'lhs = java.util.LinkedHashSet.new; lhs.map { }; puts :ok'                           
ok
```

I think the path to follow is still the same, though: package-private methods will be omitted from binding, just like private methods, from Java 17 and forward.

**Comment by headius** [comment]
This will be fixed in 9.4.6.0 via #8093.

A workaround for you is to remove the bad `map` method from LinkedHashSet:

```ruby
java.util.LinkedHashSet.remove_method(:map) rescue nil
```

The `rescue nil` will allow this to work after the patch is release, since it will raise an error if we don't bind the bad `LinkedHashSet#map` method:

```
$ jruby -e 'java.util.LinkedHashSet.remove_method(:map)'                                                   
NameError: method 'map' not defined in Java::JavaUtil::LinkedHashSet
  remove_method at org/jruby/RubyModule.java:3743
         <main> at -e:1
$ jruby -J--add-opens -Jjava.base/java.util=org.jruby.dist -e 'java.util.LinkedHashSet.remove_method(:map)'                                     
$
```
